src/NServiceBus.Core/Sagas/SagaMessage.cs

```diff
@@ -10,16 +12,22 @@ public class SagaMessage
         /// </summary>
         /// <param name="messageType">Type of the message.</param>
         /// <param name="isAllowedToStart"><code>true</code> if the message can start the saga, <code>false</code> otherwise.</param>
-        public SagaMessage(string messageType, bool isAllowedToStart)
+        public SagaMessage(Type messageType, bool isAllowedToStart)
         {
             MessageType = messageType;
+            MessageTypeName = messageType.FullName;
             IsAllowedToStartSaga = isAllowedToStart;
         }
 
         /// <summary>
         /// The type of the message.
         /// </summary>
-        public string MessageType { get; private set; }
+        public Type MessageType { get; private set; }
```

Comment:
Agree the ctor doesn't need to be public since the saga message is owned by the metadata thingy (don't remember the exact name). :+1: for making the ctor internal. I think you laid out all arguments why we can go with Type instead of string so I'll withdraw my objections